mavgen_c.py: fix trimmed message decode

jnippula opened the pull request:
This is related to Issue I have created: https://github.com/ArduPilot/pymavlink/issues/684

In case of byte swap or alignment fields, the 'mavlink_msg_xxx_get_' functions
called by message decode functions shall return actual message field value
only if the field is included in the received message. In case the field is
trimmed out it should return zero.

Example output:
```
static inline uint16_t mavlink_msg_hil_gps_get_yaw(const mavlink_message_t* msg)
{
    if (msg->len > 37)
        return _MAV_RETURN_uint16_t(msg,  37);
    return (uint16_t) 0;
}
```

JonasVautherin reviewed (commented):
This seems like it makes sense to me, but I'm not super familiar with the codebase.

One question: is it correct that `${wire_offset}` is the position of the field `${name}` in the message? Seems like it, but just to be sure :blush:

jnippula commented:
> One question: is it correct that `${wire_offset}` is the position of the field `${name}` in the message? Seems like it, but just to be sure 😊

Yes, it is.
Just realized that this could be optimized a bit by adding that length check only if the message type contains v2 extension fields (MAVLINK_MSG_ID_XXX_LEN > MAVLINK_MSG_ID_XXX_MAX_LEN).

jnippula commented:
Checked also if there would be easy way to modify empy template script a bit to add the length check only in case the message really has extension fields. However, I couldn't find quick solution for that, would require more changes in python generator functions, maybe not worth to optimize.

peterbarker commented on generator/mavgen_c.py:
```diff
@@ -384,7 +384,9 @@ def generate_message_h(directory, m):
 static inline ${return_type} mavlink_msg_${name_lower}_get_${name}(const mavlink_message_t* msg${get_arg})
 {
-    return _MAV_RETURN_${type}${array_tag}(msg, ${array_return_arg} ${wire_offset});
+    if (msg->len > ${wire_offset})
+        return _MAV_RETURN_${type}${array_tag}(msg, ${array_return_arg} ${wire_offset});
+    return (${return_type}) 0;
```
```suggestion
    if (msg->len <= ${wire_offset}) {
        // zero-bytes have been truncated from this message
        return (${return_type}) 0;
    }
    return _MAV_RETURN_${type}${array_tag}(msg, ${array_return_arg} ${wire_offset});
```

The braces because that's the style used throughout this file, even on single-line bodies - and that many bugs have been introduced by not having them :-)

I reordered to do early return and to make a nice place to put an extra comment about what's going on.

jnippula replied:
I agree, that is better solution. Patch updated.

peterbarker reviewed (changes requested):
Really good find.

peterbarker commented on generator/mavgen_c.py:
```diff
@@ -384,6 +384,10 @@ def generate_message_h(directory, m):
  */
 static inline ${return_type} mavlink_msg_${name_lower}_get_${name}(const mavlink_message_t* msg${get_arg})
 {
+    if (msg->len <= ${wire_offset}) {
```
I wonder if the spec says that if you trim some zeroes then you must trim them all.  My guess is that it doesn't.

I thought this was previously checking if the field offset was past the end of the message?

jnippula replied:
I'm not sure if I understood your point/question here..
Yes, I think nothing prevents the sender to trim only some of the zero bytes in the end, instead of trimming all of them.

The trimming process reduces the actual message length accordingly, so the receiver then just append the remaining missing zeros into the end of the received message in case the length of the received message is less than what is defined/specified for the particular message type.
Do you think there is still something missing/wrong?

peterbarker replied:
Apologies for the slow response.

My concern is that the check you have here doesn't account for the length of the field.  The wire offset may be within the length but the wire-offset-plus-field-length may be outside.  I was thinking in terms of arrays, but I guess anything longer than a byte might be a problem.

jnippula replied:
You are absolutely right. It does not work well in case of longer field types that have zeros in latter bytes (MSB or LSB depending on endianess)

jnippula replied:
Right, this is a bit tricky, implementing partial copy for all possible types and arrays would not sound very handy..

I can think of some more options:

1. Leave the mavlink_msg_<name>_get_<name> function untouched and do an intermediate copy from original msg to full length zero initialized tmp_msg and perform the above function call from that tmp_msg. Then the zeros are already in place and we are not reading from the end of the allocated space.
What to do if malloc is not successful (out of memory case)?

2. Define compiler flag to have possibility to disable sender trim-feature in case system has receiver with MAVLINK_NEED_BYTE_SWAP || !MAVLINK_ALIGNED_FIELDS defined. This is not bullet-proof, because the sender may not be configurable by developer of the receiver.

3. Get rid of the trim feature altogether. How much benefit it gives? At least there are not so many messages with "trimmable" extended fields and not so many bytes to be trimmed from them either. The feature decreases the amount of data transferred over some media, but causes extra load due to memcopy when byte swap not needed and possible incorrect data when byte swap needed.

4) Some new idea to fix this?

Personally I would vote for the option 3).

Commit:
generator/C: fix trimmed message decode

In case of byte swap or alignment fields, the 'mavlink_msg_xxx_get_' functions
called by message decode functions shall return actual message field value
only if the field is included in the received message. In case the field is
fully or partially trimmed out it should return zero instead of fetching data
outside of the received message data area.

jnippula commented:
Ok, I had some time to look at this again. I made another proposal for this.
It should now take all different types into account, also if they are partially or fully trimmed.
What do you think about it?

jnippula commented:
Added some missing changes from "put" functions.

Now when I run some more tests, I realized that the trimming does not affect only the message extension fields, but all the fields. There is no checking whether the payload field is extension one or not, it is trimmed if it contains zero bytes in the end. So, the trimming affects basically all the mavlink2.0 messages.

jnippula commented:
Well, not quite fixed yet. Still issues with byte arrays where length is not checked. Working on it.

jnippula commented:
..and fixed the MAV_RETURN_ macros for 1 byte length fields

jnippula commented:
I think the pymavlink test scripts shall verify also the case where "MAVLINK_ENDIAN = MAVLINK_BIG_ENDIAN". If I fetch the latest pymavlink (master branch), modify test Makefile to run MAVLINK_ENDIAN as MAVLINK_BIG_ENDIAN, the test_generator.sh manages to test 6 messages and then crash.

I made another PR to enable big_endian test case also:

jnippula commented:
At least this big endian enabled test set (https://github.com/ArduPilot/pymavlink/pull/812) runs now successfully with this new trimming fix implementation, so it should be in better shape now.
